File: SampleApps/SampleAppObjC/viewControllers/CustomSDKViewController.m
//
//  CustomSDKViewController.m
//  SampleAppObjC
//
//  Copyright © 2017-2020 BEAR SAS. All rights reserved.
//

#import "CustomSDKViewController.h"
#import <UserNotifications/UserNotifications.h>
#import "Cache.h"
#import "HistoryViewController.h"

#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@interface CustomSDKViewController ()

@property (weak, nonatomic) IBOutlet UIButton *startStopBtn;
@property (weak, nonatomic) IBOutlet UIButton *flashBtn;
@property (weak, nonatomic) IBOutlet UIButton *screenshotBtn;
@property (weak, nonatomic) IBOutlet UIView *splashScreen;
@property (weak, nonatomic) IBOutlet UIButton *historyBtn;

@end

@implementation CustomSDKViewController

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self initDefault];
    }
    return self;
}

-(void)initDefault {
    _enableHistory = NO;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.delegate = self;
    [_historyBtn setHidden:!_enableHistory];
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    if (self.navigationController != nil) {
        [self.navigationController setNavigationBarHidden:YES animated:animated];
    }
}

-(void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [_splashScreen setHidden:YES];
}

- (IBAction)dismiss {
    if (self.navigationController != nil) {
        [self.navigationController popViewControllerAnimated:YES];
    } else {
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}

- (IBAction)startStopTapped {
    switch (self.handler.viewState) {
        case ARViewStateIdle:
            [self.handler startScanning];
            break;
        case ARViewStateScanning:
            [self.handler stopScanning];
            break;
        default:
            break;
    }
}

- (IBAction)enableDisableFlash {
    if (!self.handler.isFlashEnabled) {
        [self.flashBtn setSelected: [self.handler enableFlash]];
    } else {
        [self.handler disableFlash];
        [_flashBtn setSelected:NO];
    }
}

- (IBAction)showARSceneWithoutTracking {
    [self.handler showARSceneWithoutTrackingWithMarkerId:226620];
}

- (IBAction)takeScreenshot {
    UIImage* image = [self.handler takeScreenshot];
    if (image != nil) {
        [self presentShareWithScreenshot:image];
    }
}

- (IBAction)cleanView {
    [self.handler cleanView];
}

- (IBAction)registerNotifications {
    
    [[UNUserNotificationCenter currentNotificationCenter]
     requestAuthorizationWithOptions:UNAuthorizationOptionAlert | UNAuthorizationOptionBadge | UNAuthorizationOptionSound
     completionHandler:^(BOOL granted, NSError * _Nullable error) {
        if (!granted) return;
        dispatch_async(dispatch_get_main_queue(), ^{
            [[UIApplication sharedApplication] registerForRemoteNotifications];
        });
    }];
}

-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if (![[segue destinationViewController] isKindOfClass:[HistoryViewController class]]) return;
    HistoryViewController* vc = (HistoryViewController*)[segue destinationViewController];
    vc.hideUnloadBtn = YES;
}

-(void)presentShareWithScreenshot:(UIImage*)screenshot {
    UIActivityViewController* activityVC = [[UIActivityViewController alloc] initWithActivityItems:@[screenshot] applicationActivities:nil];
    
    activityVC.modalPresentationStyle = UIModalPresentationPopover;
    if (activityVC.popoverPresentationController != nil) {
        activityVC.popoverPresentationController.sourceRect = _screenshotBtn.frame;
        activityVC.popoverPresentationController.sourceView = self.view;
    }
    [self presentViewController:activityVC animated:true completion:nil];
}

// MARK: ARDelegate

- (void)recognizedMarkerWithId:(NSInteger)markerId assetIds:(NSArray<NSNumber *> * _Nonnull)assetIds {
    NSLog(@"recognizedMarkerWithId = %ld, assetIds = %@", (long)markerId, assetIds);
    [Cache cacheRecognizedMarkerId:markerId];
}

- (void)recognitionTimeoutReached {
    NSLog(@"recognitionTimeoutReached");
}

- (void)assetClickedWith:(NSInteger)assetId {
    NSLog(@"assetClickedWith assetId - %li", (long)assetId);
}

- (void)viewStateChanged:(enum ARViewState)state {
    switch (state) {
        case ARViewStateIdle:
            NSLog(@"ARViewStateIdle");
            [_flashBtn setSelected:NO];
            [_startStopBtn setSelected:NO];
            break;
        case ARViewStateScanning:
            NSLog(@"ARViewStateScanning");
            [_startStopBtn setSelected:YES];
            break;
        case ARViewStateProcessing:
            NSLog(@"ARViewStateProcessing");
            [_flashBtn setSelected:NO];
            break;
        case ARViewStateTracking:
            NSLog(@"ARViewStateTracking");
            [_startStopBtn setSelected:NO];
            break;
    }
}

- (void)didFailWithError:(BearError* _Nonnull)error {
    NSLog(@"didFailWithError %@", error);
    UIAlertController* alertVC = [UIAlertController alertControllerWithTitle:error.shortDescription
                                                                     message:error.reason preferredStyle:UIAlertControllerStyleAlert];
    [alertVC addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alertVC animated:YES completion:nil];
}

@end

File: SampleApps/SampleApp/viewControllers/CustomSDKViewController.swift
//
//  CustomSDKViewController.swift
//  SampleApp
//
//  Copyright © 2017-2020 BEAR SAS. All rights reserved.
//

import Foundation
import BearSDK
import UserNotifications

final class CustomSDKViewController: ARViewController {
    
    @IBOutlet weak var startStopBtn: UIButton!
    @IBOutlet weak var flashBtn: UIButton!
    @IBOutlet weak var screenshotBtn: UIButton!
    @IBOutlet weak var splashScreen: UIView!
    @IBOutlet weak var historyBtn: UIButton!
    
    var enableHistory: Bool = false
    var hideNavBar: Bool = true
    
    override func viewDidLoad() {
        super.viewDidLoad()
        delegate = self
        historyBtn.isHidden = !enableHistory
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        if hideNavBar {
            navigationController?.setNavigationBarHidden(true, animated: true)
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        splashScreen.isHidden = true
    }
    
    @IBAction func dismiss() {
        if let nav = navigationController {
            nav.popViewController(animated: true)
        } else {
            dismiss(animated: true, completion: nil)
        }
    }
    
    @IBAction func startStopTapped() {
        switch handler.viewState {
        case .idle:
            handler.startScanning()
        case .scanning:
            handler.stopScanning()
        default:
            return
        }
    }
    
    @IBAction func enableDisableFlash() {
        if !handler.isFlashEnabled {
            flashBtn.isSelected = handler.enableFlash()
        } else { 
            handler.disableFlash()
            flashBtn.isSelected = false
        }
    }
    
    @IBAction func showARSceneWithoutTracking() {
        handler.showARSceneWithoutTracking(withMarkerId: 226620)
    }
    
    @IBAction func takeScreenshot() {
        guard let screenshot = handler.takeScreenshot else {
            print("Failed to take screenshot")
            return
        }
        presentShare(withScreenshot: screenshot)
    }
    
    @IBAction func cleanView() {
        handler.cleanView()
    }
    
    @IBAction func registerNotifications() {
        UNUserNotificationCenter.current()
            .requestAuthorization(options: [.badge, .alert , .sound]) { (granted, error) in
                guard granted else { return }
                DispatchQueue.main.async {
                    UIApplication.shared.registerForRemoteNotifications()
                }
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let vc = segue.destination as? HistoryViewController {
            vc.hideUnloadBtn = true
        }
    }
    
    private func presentShare(withScreenshot screenshot: UIImage) {
        
        let activityVC = UIActivityViewController(activityItems: [screenshot], applicationActivities: nil)
        
        activityVC.modalPresentationStyle = .popover
        activityVC.popoverPresentationController?.sourceRect = screenshotBtn.frame
        activityVC.popoverPresentationController?.sourceView = view
        
        present(activityVC, animated: true, completion: nil)
    }
}

extension CustomSDKViewController: ARDelegate {
    
    func recognizedMarker(withId markerId: Int, assetIds: [Int]) {
        print("called \(#function), markerId = \(markerId), assetIds = \(assetIds)")
        cacheRecognized(markerId: markerId)
    }
    
    func recognitionTimeoutReached() {
        print("called \(#function)")
    }
    
    func assetClicked(with assetId: Int) {
        print("called \(#function), assetId = \(assetId)")
    }
    
    func viewStateChanged(_ state: ARViewState) {
        switch state {
        case .idle:
            print("idle")
            flashBtn.isSelected = false
            startStopBtn.isSelected = false
        case .scanning:
            print("scanning")
            startStopBtn.isSelected = true
        case .processing:
            print("processing")
            flashBtn.isSelected = false
        case .tracking:
            print("tracking")
            startStopBtn.isSelected = false
        }
    }
    
    func didFail(withError error: BearError) {
        print("called \(#function)")
        showAlert(withTitle: error.shortDescription, message: error.reason)
    }
    
    private func showAlert(withTitle title: String, message: String?) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        present(alert, animated: true, completion: nil)
    }
}

